### PR TITLE
Refactoring and tests.

### LINKED ISSUE #37: Minor refactoring
Code refactoring:

1. Unwrap the module from the main function. Allow passing options needed by each task directly to the task function.
2. The module should expose a single `build()` task which will do everything from compiling to bundling, minifying, logging output. In order to compose multiple sub-tasks (which can be promise or stream based) see https://github.com/gulpjs/undertaker/issues/67.
3. Use deferred `require()` calls in order to speed up gulp bootstrap.

Functionality:

1. Currently we only support `--config` based on which the entry point for rollup is built. But it also should be possible to pass a path to an entry file.
2. There are 3 type of options:
  * The current `editor` and `plugins` options are needed to generate the entry point.
  * Paths:
    * `path` – it should be `destinationPath` and it should be relative to cwd
    * source paths:
      * `entryPoint` – path to the rollup entry point
      * or [`sourcePath` (the compilation output directory, but it doesn't have to be passed to the task, because the task will be compiling the code itself)] + `plugins` (the full paths – e.g. `autoformat/autoformat`) + `editor`
  * And the rest are rollup options (and can be defined as `rollupOptions` object, so we don't block people from using more options than we needed so far):
    * `moduleName`
    * `format`
    * ... etc
3. Add option `--no-minify`.
4. Move the default config from `ckeditor5/dev/.../...` to the root dir. Call it simply `build-config.js`. Also, we should make the `--config` option default to `./build-config.js`.

Example usage:

```
gulp build --config=./build-config.js
gulp build --config=./build-config.js --no-minify
```

Example config:

```js
'use strict';

module.exports = {
	// Specify path where bundled editor will be saved.
	destinationPath: './build/dist/',

	// The editor class and list of plugins.
	editor: 'editor-classic/classic',

	plugins: [
		'autoformat/autoformat',
		'basic-styles/bold',
		'basic-styles/italic',
		'clipboard/clipboard',
		...
	],

	rollupOptions: {
		// Name of CKEditor instance exposed as global variable by a bundle.
		moduleName: 'ClassicEditor',

		// The format of the generated bundle.
		format: 'iife'
	}
};
```

In the `gulpfile.js` I imagine to see something like:

```js
gulp.task( 'build', () => {
	const bundler = require( '@ckeditor/ckeditor5-dev-bundler-rollup' );

	return bundler.tasks.build( getBuildOptions() );
} );

function getBuildOptions() {
	const bundler = require( '@ckeditor/ckeditor5-dev-bundler-rollup' );
	const args = bundler.utils.parseArguments();

	return {
		// Needed by the compiler.
		packages: getCKEditor5PackagesPaths(),

		buildConfig: require( args.config ),

		noMinify: args.noMinify
	};
}
```

### PR DESCRIPTION
**Pull request by ma2ciek**
Fixes #37
Requires ckeditor5 with changed gulpfile to work. [t/374](https://github.com/ckeditor/ckeditor5/tree/t/374)

### CODE REVIEW

File: lib/tasks.js [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+const gutil = require( 'gulp-util' );
+const mkdirp = require( 'mkdirp' );
+const path = require( 'path' );
+const rollupBabel = require( 'rollup-plugin-babel' );
+const utils = require( './utils' );
+const compiler = require( '@ckeditor/ckeditor5-dev-compiler' );
+const streamToPromise = require( 'stream-to-promise' );
+
+module.exports = {
+	/**
```
Comment (Reinmar) [inline lib/tasks.js]: Missing docs.

File: lib/tasks.js [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+			.then( () => this.clean( destinationPath, fileName ) )
+			.then( () => this.compile( sourceBuildDir, packages ) )
+			.then( () => generate( buildConfig,	sourceBuildDir,	fileName ) )
+			.then( () => this.maybeCleanDir( sourceBuildDir ) )
+			.then( () => this.minify( destinationPath, fileName ) )
+			.then( () => this.showSummary( destinationPath, fileName ) );
+	},
+
+	/**
+	 * Remove all files from bundle directory.
```
Comment (Reinmar) [inline lib/tasks.js]: s/Remove/Removes

File: lib/tasks.js [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+		const { destinationPath, entryPoint } = buildConfig;
+
+		const generate = entryPoint ?
+			this.generateBundleWithEntryPoint.bind( this ) :
+			this.generateBundleWithoutEntryPoint.bind( this );
+
+		return Promise.resolve()
+			.then( () => this.clean( destinationPath, fileName ) )
+			.then( () => this.compile( sourceBuildDir, packages ) )
+			.then( () => generate( buildConfig,	sourceBuildDir,	fileName ) )
```
Comment (Reinmar) [inline lib/tasks.js]: Mixed tabs with spaces (there should be no tabs in an arguments list)

File: lib/tasks.js [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+				esnext: sourceBuildDir,
+			},
+			packages,
+		} );
+	},
+
+	/**
+	 * Combine whole editor files into two files `ckeditor.js` and `ckeditor.css`.
+	 *
+	 * ** Note: ** Entry file is a physically existing file because rollup requires it.
```
Comment (Reinmar) [inline lib/tasks.js]: Unnecessary spaces inside `** **`

File: lib/tasks.js [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+		// Create a temporary entry file with proper directory structure if not exist.
+		const {
+			rollupOptions,
+			destinationPath,
+			entryPoint,
+		} = buildConfig;
+
+		const filePath = path.join( destinationPath, fileName );
+
+		// Lets wait for both - JS and CSS.
```
Comment (Reinmar) [inline lib/tasks.js]: Unnecessary comment.

File: lib/tasks.js (hunk 4 above)
Comment (Reinmar) [inline lib/tasks.js]: This repeats couple of times in this PR.

File: lib/tasks.js [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+			this.generateCssBundle( {
+				sourceBuildDir,
+				fileName,
+				filePath
+			} ),
+		] );
+	},
+
+	/**
+	 * Create entry file
```
Comment (Reinmar) [inline lib/tasks.js]: Missing period.

File: lib/tasks.js [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+	/**
+	 * Bundle JS by Rollup.
+	 *
+	 * @param {Object}
+	 * @param {String} entryPoint
+	 * @param {Object} rollupOptions
+	 * @param {String} rollupOptions.moduleName Module name of CKEditor instance exposed as global variable by a bundle.
+	 * @param {String} bundleTmpDir
+	 */
+	generateJsBundle( { entryPoint, filePath, rollupOptions, bundleTmpDir } ) {
```
Comment (Reinmar) [inline lib/tasks.js]: What's the difference between "generate" and "create"? The names are really confusing.

File: lib/tasks.js [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+		}, rollupOptions ) );
+	},
+
+	/**
+	 * Clean directory if the path to directory is provided
+	 *
+	 * @param {String} [dir]
+	 * @returns {Promise}
+	 */
+	maybeCleanDir( dir ) {
```
Comment (Reinmar) [inline lib/tasks.js]: The couple of functions there are no tasks. Clean this file by either moving them to private scope or to a separate utils file. Actually, the only task here which we want to be public now is `build()`. The rest should be protected.

File: tests/tasks.js [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+			sandbox.stub( path, 'join', ( ...args ) => args.join( '/' ) );
+
+			return tasks.minify( 'destinationPath', 'fileName' );
+		} );
+	} );
+
+	describe( 'minifyJs', () => {
+		it( 'should minify js file', () => {
+			sandbox.stub( gulp, 'src', () => ( {
+				pipe: () => { }
```
Comment (Reinmar) [inline tests/tasks.js]: `{}` (and this repeats couple of times)

File: tests/tasks.js [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+			} ) );
+			sandbox.stub( path, 'dirname', () => {} );
+			const saveMinified = sandbox.stub( utils, 'saveFileFromStreamAsMinified' );
+
+			tasks.minifyJs( 'path/to/editor' );
+
+			expect( saveMinified.called ).to.be.eq( true );
```
Comment (Reinmar) [inline tests/tasks.js]: Nothing else that we can check here? Path maybe?

**Comment by pomek** [comment]
I'm for all the options:
- mark protection methods in JSDoc (`@protected`),
- use underscore as a prefix (`_protectedMethod`),
- describe public methods in README.

File: lib/tasks.js (hunk 1 above)
Comment (pomek) [inline lib/tasks.js]: In all dev modules, we are using a little bit different convention:

```js
// Definition of tasks
const tasks = {};

module.exports = tasks;
```

It allows avoiding to use the `this` keyword in tasks. Also, we don't have to use [Function.prototype.bind](https://github.com/ckeditor/ckeditor5-dev-bundler-rollup/pull/38/files#diff-17a298a4fcdf2c8fee67743145edf5f1R37) _(link leads to the part of the code)_.

Comment (ma2ciek) [inline lib/tasks.js]: Ok, thanks for the guides. Fixed.

**Comment by ma2ciek** [comment]
@pomek I have just added `@public` and `@protected` (+prefix) tags  in code. Main function is already explained in the README.

**Comment by pomek** [comment]
I'm not sure whether to we need the `@public` annotation.

@Reinmar could you help us?

File: lib/tasks.js [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+	/**
+	 * Main function that takes packages and build config and runs these tasks:
+	 * 1. Compiles project
+	 * 2. Generates js and css bundles (both normal and minified).
+	 * 3. Cleans temp dir
+	 *
+	 * To create bundle from existing entry point add path to the build config object.
+	 * Otherwise add path to editor and list of plugins to the build config plugins.
+	 *
+	 * @public
```
Comment (Reinmar) [inline lib/tasks.js]: Don't need to mark the public APi like this.

File: README.md [hunk 12]
```diff
@@ -23,41 +23,28 @@ To include development tasks in your `gulpfile.js`:
...
+}
+
+function getCKEditor5PackagesPaths() {
+	const compiler = require( '@ckeditor/ckeditor5-dev-compiler' );
+	return compiler.utils.getPackages( config.ROOT_DIR );
+}
+
 ```
 
 **Please note that `compile:*` tasks are available in [`ckeditor5-dev-compiler`](https://github.com/ckeditor/ckeditor5-dev-builder/) package.**
```
Comment (Reinmar) [inline README.md]: This note is not needed any more.

File: README.md [hunk 13]
```diff
@@ -23,41 +23,28 @@ To include development tasks in your `gulpfile.js`:
 
 ```js
 const gulp = require( 'gulp' );
-const runSequence = require( 'run-sequence' );
```
Comment (Reinmar) [inline README.md]: There's a line above "Description of each...". You can remove it now.

**Comment by Reinmar** [comment]
You removed the wrong line: https://github.com/ckeditor/ckeditor5-dev-bundler-rollup/pull/38/commits/3bf108a719bf4086e227aa10668c0f847bedef11#diff-04c6e90faac2675aa89e2176d2eec7d8L17 :D

**Comment by Reinmar** [comment]
And instead of adding a link to the issue, please list there the options and example configs.

**Comment by Reinmar** [comment]
I tried this config:

```
module.exports = {
	// Specify path where bundled editor will be saved.
	destinationPath: './build/dist/',

	entryPoint: './entry.js',

	rollupOptions: {
		// Name of CKEditor instance exposed as global variable by a bundle.
		moduleName: 'ClassicEditor',

		// The format of the generated bundle.
		format: 'iife'
	}
};
```

With this entry file:

```js
import ClassicEditor from './build/modules/esnext/ckeditor5/editor-classic/classic.js';
import Image from './build/modules/esnext/ckeditor5/image/image.js';

export default { ClassicEditor, Image };
```

And got:

```
[16:38:11] Finished writing '/www/ckeditor5/ckeditor5/build/temp-build/ckeditor5/utils/lib/lodash/zipWith.js'
[16:38:11] Error: You must supply options.entry to rollup
    at checkOptions (/www/ckeditor5/ckeditor5-dev-bundler-rollup/node_modules/rollup/src/rollup.js:46:10)
    at Object.rollup (/www/ckeditor5/ckeditor5-dev-bundler-rollup/node_modules/rollup/src/rollup.js:60:16)
    at Object._rollupBundle (/www/ckeditor5/ckeditor5-dev-bundler-rollup/lib/tasks.js:230:17)
    at Promise.resolve.then (/www/ckeditor5/ckeditor5-dev-bundler-rollup/lib/tasks.js:216:23)
    at process._tickDomainCallback (internal/process/next_tick.js:129:7)
[16:38:11] 'build' errored after 5.74 s
[16:38:11] Error: Build error.
    at Object._handleRollupError (/www/ckeditor5/ckeditor5-dev-bundler-rollup/lib/tasks.js:280:9)
    at Promise.resolve.then.then.then.catch.err (/www/ckeditor5/ckeditor5-dev-bundler-rollup/lib/tasks.js:219:25)
    at process._tickDomainCallback (internal/process/next_tick.js:129:7)
```

**Comment by Reinmar** [comment]
I found the issue. I'll work on it.

**Pull request merged**